File: LeetCode/Largest Rectangle in Histogram.py
class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
        l = len(height)
        if l==0:
            return 0
        indexStack = []
        area = [0] * l
        height.append(0)
        maxArea=0
        i=0
        while i <= l:
            if not indexStack or height[indexStack[-1]] <= height[i]:
                indexStack.append(i)
                i+=1
            else:
                idx = indexStack.pop()
                area[idx] += (i-idx) * height[idx]

        i=l-1
        indexStack = []
        while i >= -1:
            # print indexStack
            if not indexStack or height[indexStack[-1]] <= height[i]:
                indexStack.append(i)
                i-=1
            else:
                idx = indexStack.pop()
                # print idx,i
                area[idx] += (idx - i) * height[idx]
        for i in range(l):    
            area[i] -= height[i] 
        # print area
        return max(area)

if __name__ == '__main__':
    s=Solution()
    print s.largestRectangleArea([1,2,3,1,1,1,1])
    print s.largestRectangleArea( [2,1,5,6,2,3])
    print s.largestRectangleArea([4,2,3])
    

File: LeetCode/Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int> &height) {
        if (height.size() == 0)
            return 0;
        int maxArea = 0, l = height.size();
        int * row = new int[l+1];

        row[l] = 0;
        stack<int> st;
        for (int i = 0; i < l+1; ++i)
        {
            if (i < l) row[i] = height[i];

            if (st.empty() || row[st.top()] <= row[i])
                st.push(i);
            else{
                while(!st.empty() && row[st.top()] > row[i]){
                    int t = st.top();
                    st.pop();
                    int area = row[t] * ((st.empty())?i:(i - st.top() - 1));
                    maxArea = (maxArea < area)?area: maxArea;
                }
                st.push(i);
            }
        }

        return maxArea;
    }
};